README.md
# ELS - Enterprise Linux/EuroLinux Sources - rpmlint
 
## Usage:
  Checkout branch or tag. Use `make sources` to recreate  SOURCE0 archive.


src/ChangeLog
2013-05-09  Ville Skyttä <ville.skytta@iki.fi>

	* Makefile: Release 1.5.

2013-05-09  Ville Skyttä <ville.skytta@iki.fi>

	* BinariesCheck.py: Spelling fixes, wrap long lines.

2013-05-09  Ville Skyttä <ville.skytta@iki.fi>

	* TagsCheck.py: Cleanups.

2013-05-09  Ville Skyttä <ville.skytta@iki.fi>

	* FilesCheck.py: Add Python 3.4 magic number.

2013-04-03  Ville Skyttä <ville.skytta@iki.fi>

	* __isocodes__.py: Regenerate ISO codes list with iso-codes 3.41.

2013-03-16  Ville Skyttä <ville.skytta@iki.fi>

	* BinariesCheck.py: Treat Lua bytecode files as binary.  https://bugzilla.redhat.com/919869

2013-02-02  Ville Skyttä <ville.skytta@iki.fi>

	* Pkg.py, rpmdiff: Cleanups.

2013-02-02  Ville Skyttä <ville.skytta@iki.fi>

	* rpmlint: Revert "small refactoring" This reverts commit eb981e7f950dbe8db81594220e15482c3b451c4e.  The global config file does imports, sets up stuff etc for user
	config files and /etc/rpmlint/*config snippets to use, and
	execfile()ing the global one in a code block like this doesn't do
	that properly, resulting in for example addFilter() not being
	available for these additional configs.

2013-02-01  Michael Scherer <misc@zarb.org>

	* BinariesCheck.py: add a help text, as suggested by Dan Horák on
	freenode

2013-01-26  Michael Scherer <misc@zarb.org>

	* FilesCheck.py: Private keys should also not be shipped in rpm, for
	the reason explained in the commit.

2013-01-26  Michael Scherer <misc@zarb.org>

	* FilesCheck.py: warn when a pem certificate is found in a package.  after the compromise of various CA in the last years, it would be
	better to not have X copy of the CA certificate in each package.  Not to mention the issue of having some packages shipping a default
	private key with the certificate for a perfect out of box
	experience, that could lead to security issue. So the check look for
	certificate, but verify that the file is not not marked as %ghost ).  idea from kiilerix on irc

2013-01-19  Michael Scherer <misc@zarb.org>

	* rpmlint: small refactoring

2013-01-19  Michael Scherer <misc@zarb.org>

	* FilesCheck.py: use a more specific exception ( as this could mask
	others issues )

2013-01-18  Michael Scherer <misc@zarb.org>

	* MenuXDGCheck.py: add desktopfile-without-binary warning, to check
	if the file point to a existing binary ( idea originally from Ed
	Santiago )

2013-01-18  Michael Scherer <misc@zarb.org>

	* FilesCheck.py: refactor a serie of if/elif into a single loop

2013-01-18  Michael Scherer <misc@zarb.org>

	* FilesCheck.py: shorten the regexp for /tmp directory

2013-01-18  Michael Scherer <misc@zarb.org>

	* SpecCheck.py: check that the packager didn't get sloppy by using
	_default_patch_fuzz, idea originally from Ed Santiago

2013-01-18  Michael Scherer <misc@zarb.org>

	* SpecCheck.py: refactor the regexp code for define and global (
	easier to read, less risk of errrors )

2013-01-17  Michael Scherer <misc@zarb.org>

	* Common.py, PostCheck.py, SpecCheck.py: start to refactor various
	copy of tag ( since %verifyscript was missing in SpecCheck.py )

2013-01-17  Michael Scherer <misc@zarb.org>

	* SpecCheck.py: refactor regexp used for the tag matching

2013-01-17  Michael Scherer <misc@zarb.org>

	* FHSCheck.py: removed /usr/spool and /usr/X386, reorder /var
	subdir, removed /var/state, added /var/mail, to sync with latest FHS

2013-01-17  Michael Scherer <misc@zarb.org>

	* PostCheck.py: small refactoring ( could go further )

2013-01-15  Michael Scherer <misc@zarb.org>

	* FilesCheck.py: add /run, kinda standard nowadays

2013-01-15  Michael Scherer <misc@zarb.org>

	* BinariesCheck.py: refactor the code of regexp into a common
	function, easier to maintain and to read

2013-01-13  Michael Scherer <misc@zarb.org>

	* BinariesCheck.py: add a check against insecure usage of mktemp
	function Since the function is insecure on some platform, and suffer from a
	race condition on every platform, it should not be used ( as
	explained in the manpage ). The function was also removed from
	POSIX.1-2008, according to the manpage too, and call should better
	be replaced by mkstemp.

2013-01-13  Michael Scherer <misc@zarb.org>

	* BinariesCheck.py: removed the regexp in a attempt to speed up the
	test Unfortunately, this didn't changed much but 1s is still a small win.

2013-01-13  Michael Scherer <misc@zarb.org>

	* BinariesCheck.py: check that the chdir call is nor chroot ( ie 2
	functions call near ).  yet the function is quite slow, and should be done without the
	regexp, since this slow down a lot for a big binary.

2013-01-13  Michael Scherer <misc@zarb.org>

	* BinariesCheck.py: add a first preliminary port of
	http://people.redhat.com/sgrubb/security/find-chroot for now, it just check if chdir and chroot are present in compiled
	software, and the next commit should verify that the call to chdir
	is near the one to chroot.

2013-01-12  Michael Scherer <misc@zarb.org>

	* BinariesCheck.py: Add a check stolen from steve grubb security
	page, to detect if a binary use setuid and setgid without reseting
	supplementary group using initgroups, or setgroups. Binaries who do
	this should be fixed, as this could be used for privilege escalation
	( for example, by using 'adm' or 'wheel' group ).

2012-12-21  Ville Skyttä <ville.skytta@iki.fi>

	* FilesCheck.py: Link to Python files containing the list of magic
	values.

2012-11-05  Ville Skyttä <ville.skytta@iki.fi>

	* INSTALL: Note groff dependency.

2012-09-24  Ville Skyttä <ville.skytta@iki.fi>

	* FilesCheck.py: Treat RDoc RI files as binary.  Based on patch by Josef Strzibny
	<strzibny@dhcp-25-242.brq.redhat.com>,
	http://rpmlint.zarb.org/cgi-bin/trac.cgi/ticket/569

2012-08-23  Ville Skyttä <ville.skytta@iki.fi>

	* AbstractCheck.py: Tighten macro regexp to min 3 chars, starting
	with a letter or underscore.  Based on patch by Josef Strzibny
	<strzibny@dhcp-25-242.brq.redhat.com>,
	http://rpmlint.zarb.org/cgi-bin/trac.cgi/ticket/569

2012-07-13  Ville Skyttä <ville.skytta@iki.fi>


src/Config.py
# -*- coding: utf-8 -*-
#############################################################################
# File          : Config.py
# Package       : rpmlint
# Author        : Frederic Lepied
# Created on    : Fri Oct 15 20:04:25 1999
# Version       : $Id: Config.py 1711 2010-01-21 21:57:11Z scop $
# Purpose       : handle configuration options. To be used from config files.
#############################################################################

import locale
import os.path
import re

try:
    from __version__ import __version__
except ImportError:
    __version__ = 'devel'

DEFAULT_CHECKS = ("DistributionCheck",
                  "TagsCheck",
                  "BinariesCheck",
                  "ConfigCheck",
                  "FilesCheck",
                  "DocFilesCheck",
                  "FHSCheck",
                  "SignatureCheck",
                  "I18NCheck",
                  "MenuCheck",
                  "PostCheck",
                  "InitScriptCheck",
                  "SourceCheck",
                  "SpecCheck",
                  "NamingPolicyCheck",
                  "ZipCheck",
                  "PamCheck",
                  "RpmFileCheck",
                  "MenuXDGCheck",
                  )

USEUTF8_DEFAULT = False
try:
    if locale.getpreferredencoding() == 'UTF-8':
        USEUTF8_DEFAULT = True
except:
    try:
        if re.match('utf', locale.getdefaultlocale()[1], re.I):
            USEUTF8_DEFAULT = True
    except:
        pass

info = False
no_exception = False

# handle the list of checks to load
_checks = []
_checks.extend(DEFAULT_CHECKS)

def addCheck(check):
    global _checks

    check = re.sub('\.py[co]?$', '', check)
    if check not in _checks:
        _checks.append(check)

def allChecks():
    global _checks

    if _checks == []:
        defaultChecks()
    return _checks

def defaultChecks():
    resetChecks()
    _checks.extend(DEFAULT_CHECKS)

def resetChecks():
    global _checks

    _checks = []

# handle the list of directories to look for checks

_dirs = ["/usr/share/rpmlint"]

def addCheckDir(dir):
    global _dirs

    d = os.path.expanduser(dir)

    if d not in _dirs:
        _dirs.insert(0, d)

def checkDirs():
    global _dirs

    return _dirs

# handle options

_options = {}

def setOption(name, value):
    global _options

    _options[name] = value

def getOption(name, default = ""):
    global _options

    try:
        return _options[name]
    except:
        return default

# List of filters
_filters = []
_filters_re = None

def addFilter(s):
    global _filters
    global _filters_re

    _filters.append(s)
    _filters_re = None

def removeFilter(s):
    global _filters
    global _filters_re

    try:
        _filters.remove(s)
    except:
        pass
    else:
        _filters_re = None

_scoring = {}

def setBadness(s, score):
    _scoring[s] = score

def badness(s):
    return _scoring.get(s, 0)

_non_named_group_re = re.compile('[^\\](\()[^:]')
def isFiltered(s):
    global _filters
    global _filters_re

    if _filters_re == None:
        # no filter
        if len(_filters) == 0:
            return False
        _filters_re = '(?:' + _filters[0] + ')'

        for idx in range(1, len(_filters)):
            # to prevent named group overflow that happen when there is too many () in a single regexp
            # AssertionError: sorry, but this version only supports 100 named groups
            if '(' in _filters[idx]:
                _non_named_group_re.subn('(:?', _filters[idx])
            _filters_re = _filters_re + '|(?:' + _filters[idx] +')'
        _filters_re = re.compile(_filters_re)

    if not no_exception:
        if _filters_re.search(s):
            return True
    return False

# Config.py ends here

# Local variables:
# indent-tabs-mode: nil
# py-indent-offset: 4
# End:
# ex: ts=4 sw=4 et
